Userland/Libraries/LibGfx/BitmapSequence.cpp

```diff
@@ -0,0 +1,129 @@
+            TRY(encoder.encode(bitmap->size_in_bytes()));
+            TRY(encoder.encode(bitmap->size()));
+
+            total_buffer_size += bitmap_option.value()->size_in_bytes();
+        }
+    }
+
+    // collate all of the bitmap data into one contiguous buffer
+    auto collated_buffer = TRY(Core::AnonymousBuffer::create_with_size(total_buffer_size));
+
+    auto* write_pointer = collated_buffer.data<u8>();
+    for (auto const& bitmap_option : bitmaps) {
+        if (bitmap_option.has_value()) {
+            auto bitmap = bitmap_option.value();
+            memcpy(write_pointer, bitmap->scanline(0), bitmap->size_in_bytes());
+            write_pointer += bitmap->size_in_bytes();
+        }
+    }
+
+    TRY(encoder.encode(collated_buffer));
+
+    return {};
+}
+
+template<>
+ErrorOr<Gfx::BitmapSequence> decode(Decoder& decoder)
+{
+    // a struct to temporarily store bitmap fields before the buffer data is decoded
+    struct BitmapMetadata {
+        Gfx::BitmapFormat format;
+        Gfx::AlphaType alpha_type;
+        Gfx::IntSize size;
+        size_t size_in_bytes;
+    };
+
+    Vector<Optional<BitmapMetadata>> bitmaps_metadata;
+
+    size_t num_bitmaps = TRY(decoder.decode_size());
+
+    for (size_t i = 0; i < num_bitmaps; i++) {
+        auto valid = TRY(decoder.decode<bool>());
+
+        if (valid) {
+            auto raw_bitmap_format = TRY(decoder.decode<u32>());
+            if (!Gfx::is_valid_bitmap_format(raw_bitmap_format))
+                return Error::from_string_literal("IPC: Invalid Gfx::BitmapSequence format");
+            auto format = static_cast<Gfx::BitmapFormat>(raw_bitmap_format);
+
+            auto raw_alpha_type = TRY(decoder.decode<u32>());
+            if (!Gfx::is_valid_alpha_type(raw_alpha_type))
+                return Error::from_string_literal("IPC: Invalid Gfx::BitmapSequence alpha type");
+            auto alpha_type = static_cast<Gfx::AlphaType>(raw_alpha_type);
+
+            auto size_in_bytes = TRY(decoder.decode<size_t>());
+            auto size = TRY(decoder.decode<Gfx::IntSize>());
+
+            bitmaps_metadata.append(BitmapMetadata { format, alpha_type, size, size_in_bytes });
+        } else {
+            bitmaps_metadata.append({});
+        }
+    }
+
+    auto collated_buffer = TRY(decoder.decode<Core::AnonymousBuffer>());
+    auto* read_pointer = collated_buffer.data<u8>();
+
+    Vector<Optional<NonnullRefPtr<Gfx::Bitmap>>> bitmaps;
+
+    // sequentially read each valid bitmap's data from the collated buffer
+    for (auto metadata_option : bitmaps_metadata) {
+        if (metadata_option.has_value()) {
+            auto metadata = metadata_option.value();
+            size_t size_in_bytes = metadata.size_in_bytes;
+
+            auto buffer = TRY(Core::AnonymousBuffer::create_with_size(size_in_bytes));
+            memcpy(buffer.data<u8>(), read_pointer, size_in_bytes);
+            read_pointer += size_in_bytes;
```

Comment:
This part is spooky.

We must verify that the size given by the bitmap metadata makes sense. The entire transferred IPC data is "untrusted". 

So, if we want to use the size from the metadata to read from the buffer, we have to validate that size_in_bytes + read_ptr is in bounds first.

I think a better way to approach it would be to make heavy use of our AK::Span type. Specifically the type alias `ReadonlyBytes`. 

Then you can call `.offset()` or `.slice()` to get new spans that are size-checked against the parent span. you just have to make sure that a size error doesn't crash the program with an assert, but returns an AK::Error.